aspnetcore/web-api/index.md

```diff
@@ -3,14 +3,591 @@ title: Create web APIs with ASP.NET Core
+|[`[FromServices]`](xref:mvc/controllers/dependency-injection#action-injection-with-fromservices) | The request service injected as an action parameter |
+
+> [!WARNING]
+> Don't use `[FromRoute]` when values might contain `%2f` (that is `/`). `%2f` won't be unescaped to `/`. Use `[FromQuery]` if the value might contain `%2f`.
+
+Without the `[ApiController]` attribute or binding source attributes like `[FromQuery]`, the ASP.NET Core runtime attempts to use the complex object model binder. The complex object model binder pulls data from value providers in a defined order.
+
+In the following example, the `[FromQuery]` attribute indicates that the `discontinuedOnly` parameter value is provided in the request URL's query string:
+
+[!code-csharp[](index/samples/3.x/Controllers/ProductsController.cs?name=snippet_BindingSourceAttributes&highlight=3)]
+
+The `[ApiController]` attribute applies inference rules for the default data sources of action parameters. These rules save you from having to identify binding sources manually by applying attributes to the action parameters. The binding source inference rules behave as follows:
+
+* `[FromBody]` is inferred for complex type parameters not registered in the DI Container. An exception to the `[FromBody]` inference rule is any complex, built-in type with a special meaning, such as <xref:Microsoft.AspNetCore.Http.IFormCollection> and <xref:System.Threading.CancellationToken>. The binding source inference code ignores those special types.
+* `[FromForm]` is inferred for action parameters of type <xref:Microsoft.AspNetCore.Http.IFormFile> and <xref:Microsoft.AspNetCore.Http.IFormFileCollection>. It's not inferred for any simple or user-defined types.
+* `[FromRoute]` is inferred for any action parameter name matching a parameter in the route template. When more than one route matches an action parameter, any route value is considered `[FromRoute]`.
+* `[FromQuery]` is inferred for any other action parameters.
+
+### FromBody inference notes
+
+`[FromBody]` isn't inferred for simple types such as `string` or `int`. Therefore, the `[FromBody]` attribute should be used for simple types when that functionality is needed.
+
+When an action has more than one parameter bound from the request body, an exception is thrown. For example, all of the following action method signatures cause an exception:
+
+* `[FromBody]` inferred on both because they're complex types.
+
+  ```csharp
+  [HttpPost]
+  public IActionResult Action1(Product product, Order order)
+  ```
+
+* `[FromBody]` attribute on one, inferred on the other because it's a complex type.
+
+  ```csharp
+  [HttpPost]
+  public IActionResult Action2(Product product, [FromBody] Order order)
+  ```
+
+* `[FromBody]` attribute on both.
+
+  ```csharp
+  [HttpPost]
+  public IActionResult Action3([FromBody] Product product, [FromBody] Order order)
+  ```
+
+### FromServices inference notes
```

Comment:
The `FromServices` inference is not available in .NET 6. Do you think is ok add it here?

Reply:
removed